File: debug_sigsegv_getenv/repro_getenv_race.c
/*
 * Reproduction for SIGSEGV in getenv() due to glibc thread-safety bug.
 *
 * glibc's setenv() can reallocate the environ array while getenv() on another
 * thread is iterating it, causing a use-after-free / SIGSEGV.
 *
 * This is the root cause of the crash seen in Ray's OpenTelemetry metric
 * exporter thread, where gRPC calls grpc_core::GetEnv() -> getenv() on a
 * background thread while Python worker threads call os.environ[k]=v
 * (which calls setenv()).
 *
 * Build:  gcc -O2 -pthread -o repro_getenv_race repro_getenv_race.c
 * Run:    ./repro_getenv_race
 *
 * On vulnerable glibc (<2.40), this will typically SIGSEGV within seconds.
 * On patched glibc (>=2.40, e.g. Fedora 41, RHEL 10), it should run safely.
 */

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NUM_READER_THREADS 4
#define NUM_ITERATIONS 10000000

/* Names used by setenv - use many unique names to force environ array growth */
static char setenv_names[512][32];

static void *setenv_thread(void *arg) {
    (void)arg;
    char value[64];
    for (int i = 0; i < NUM_ITERATIONS; i++) {
        /* Use rotating names to force environ array to grow and reallocate */
        int idx = i % 512;
        snprintf(value, sizeof(value), "value_%d", i);
        setenv(setenv_names[idx], value, 1);
    }
    return NULL;
}

static void *getenv_thread(void *arg) {
    int thread_id = *(int *)arg;
    (void)thread_id;
    for (int i = 0; i < NUM_ITERATIONS; i++) {
        /*
         * Read various env vars, mimicking what gRPC does internally.
         * grpc_core::GetEnv() calls getenv("GRPC_EXPERIMENTS"),
         * getenv("GRPC_SHUFFLE_PICK_FIRST"), etc.
         */
        volatile const char *val;
        val = getenv("PATH");
        (void)val;
        val = getenv("HOME");
        (void)val;
        val = getenv("GRPC_EXPERIMENTS");
        (void)val;
        val = getenv("NONEXISTENT_VAR_12345");
        (void)val;
    }
    return NULL;
}

int main(void) {
    /* Pre-generate setenv key names */
    for (int i = 0; i < 512; i++) {
        snprintf(setenv_names[i], sizeof(setenv_names[i]),
                 "REPRO_ENV_%03d", i);
    }

    printf("Reproducing getenv/setenv race condition (glibc thread-safety bug)\n");
    printf("  Writer threads: 1 (calling setenv in a loop)\n");
    printf("  Reader threads: %d (calling getenv in a loop)\n", NUM_READER_THREADS);
    printf("  Iterations: %d per thread\n", NUM_ITERATIONS);
    printf("\n");
    printf("On vulnerable glibc (<2.40), expect SIGSEGV within seconds.\n");
    printf("On patched glibc (>=2.40), this should complete without crashing.\n");
    printf("\nStarting...\n");
    fflush(stdout);

    pthread_t writer;
    pthread_t readers[NUM_READER_THREADS];
    int reader_ids[NUM_READER_THREADS];

    /* Start all threads */
    pthread_create(&writer, NULL, setenv_thread, NULL);
    for (int i = 0; i < NUM_READER_THREADS; i++) {
        reader_ids[i] = i;
        pthread_create(&readers[i], NULL, getenv_thread, &reader_ids[i]);
    }

    /* Wait for completion (if we get here without SIGSEGV, the glibc is safe) */
    pthread_join(writer, NULL);
    for (int i = 0; i < NUM_READER_THREADS; i++) {
        pthread_join(readers[i], NULL);
    }

    printf("\nCompleted without crash - your glibc appears to be safe.\n");
    return 0;
}

File: debug_sigsegv_getenv/repro_getenv_race.py
"""
Reproduction for SIGSEGV in getenv() in the context of Ray + gRPC + OpenTelemetry.

This script mimics the exact race condition that causes the crash:
  - Thread A: Mutates os.environ (like Ray worker threads setting
    CUDA_VISIBLE_DEVICES, OMP_NUM_THREADS, runtime env vars, etc.)
  - Thread B: Makes gRPC calls that trigger internal getenv() via
    grpc_core::GetEnv() -> ShufflePickFirstEnabled() -> getenv()

The gRPC call triggers DNS resolution and load-balancing config parsing,
which lazily calls getenv("GRPC_EXPERIMENTS") on the gRPC thread.
Meanwhile the Python thread is calling setenv() via os.environ[k] = v.

On vulnerable glibc (<2.40), this will typically SIGSEGV within seconds.

Usage:
    pip install grpcio
    python repro_getenv_race.py
"""

import os
import sys
import threading
import time


def env_writer_thread(stop_event: threading.Event):
    """
    Simulates Ray worker threads that modify os.environ during task execution.

    In Ray, these mutations happen in:
      - set_omp_num_threads()           -> os.environ["OMP_NUM_THREADS"] = ...
      - set_visible_accelerator_ids()   -> os.environ["CUDA_VISIBLE_DEVICES"] = ...
      - update_envs()                   -> os.environ[key] = value  (runtime env)
    """
    i = 0
    while not stop_event.is_set():
        # Rotate through many unique keys to force environ array growth/reallocation
        key = f"RAY_REPRO_ENV_{i % 200}"
        os.environ[key] = f"value_{i}"
        i += 1
    print(f"  Writer thread did {i} setenv() calls")


def grpc_caller_thread(stop_event: threading.Event):
    """
    Simulates the OpenTelemetry PeriodicExportingMetricReader thread that
    exports metrics via gRPC.

    The actual crash path is:
      OtlpGrpcMetricExporter::Export()
        -> gRPC BlockingUnaryCallImpl
          -> DNS resolution (c-ares)
            -> PickFirstFactory::ParseLoadBalancingConfig()
              -> ShufflePickFirstEnabled()
                -> grpc_core::GetEnv()
                  -> getenv()  <-- SIGSEGV

    We trigger this by making gRPC calls to a non-existent endpoint,
    which forces gRPC to go through DNS resolution and LB config parsing.
    """
    try:
        import grpc
    except ImportError:
        print("ERROR: grpcio not installed. Run: pip install grpcio")
        sys.exit(1)

    calls = 0
    while not stop_event.is_set():
        try:
            # Create a channel to a non-existent endpoint.
            # This forces gRPC to perform DNS resolution, which triggers
            # the lazy getenv() call in ShufflePickFirstEnabled().
            channel = grpc.insecure_channel(
                f"nonexistent-host-{calls % 50}.example.com:50051"
            )
            # Force the channel to connect (triggers DNS + LB config parsing)
            try:
                grpc.channel_ready_future(channel).result(timeout=0.01)
            except grpc.FutureTimeoutError:
                pass
            channel.close()
            calls += 1
        except Exception as e:
            # Ignore connection errors - we only care about the getenv race
            calls += 1

    print(f"  gRPC thread did {calls} connection attempts")


def main():
    duration = 30  # seconds

    print("Reproducing getenv/setenv race condition (glibc thread-safety bug)")
    print()
    print("This mimics the Ray crash where:")
    print("  Thread A: Python worker sets os.environ (-> glibc setenv())")
    print("  Thread B: gRPC/OpenTelemetry metric export calls getenv() internally")
    print()
    print(f"Running for {duration} seconds...")
    print(f"On vulnerable glibc (<2.40), expect SIGSEGV (exit code -11).")
    print(f"On patched glibc (>=2.40), this should complete without crashing.")
    print()
    sys.stdout.flush()

    stop_event = threading.Event()

    # Start multiple writer threads (simulating Ray worker env mutations)
    writers = []
    for i in range(3):
        t = threading.Thread(target=env_writer_thread, args=(stop_event,), daemon=True)
        t.start()
        writers.append(t)

    # Start multiple gRPC threads (simulating OpenTelemetry metric export)
    grpc_threads = []
    for i in range(2):
        t = threading.Thread(target=grpc_caller_thread, args=(stop_event,), daemon=True)
        t.start()
        grpc_threads.append(t)

    # Let it run
    time.sleep(duration)

    print(f"\nStopping after {duration}s...")
    stop_event.set()

    for t in writers + grpc_threads:
        t.join(timeout=5)

    # Clean up the env vars we created
    for i in range(200):
        os.environ.pop(f"RAY_REPRO_ENV_{i}", None)

    print("\nCompleted without crash - your glibc appears to be safe.\n")


if __name__ == "__main__":
    main()
